kotlin/src/main/com/looker/rtl/OAuthSession.kt

```diff
@@ -0,0 +1,189 @@
+}
+
+fun hashString(input: ByteArray, type: String): String {
+    val digester = MessageDigest.getInstance(type)
+    return hashString(input, digester)
+}
+
+/**
+ * Supported algorithms on Android:
+ *
+ * Algorithm	Supported API Levels
+ * MD5          1+
+ * SHA-1	    1+
+ * SHA-224	    1-8,22+
+ * SHA-256	    1+
+ * SHA-384	    1+
+ * SHA-512	    1+
+ */
+fun hashString(input: String, type: String): String {
+    return hashString(input.toByteArray(), type)
+}
+
+private val hexArray = "0123456789abcdef".toCharArray()
+
+fun hexStr(bytes: ByteArray): String {
+    val hexChars = CharArray(bytes.size * 2)
+    for (j in bytes.indices) {
+        val v = (bytes[j] and 0xFF.toByte()).toInt()
+
+        hexChars[j * 2] = hexArray[v ushr 4]
+        hexChars[j * 2 + 1] = hexArray[v and 0x0F]
+    }
+    return String(hexChars)
+}
+
+@ExperimentalUnsignedTypes
+class OAuthSession(override val apiSettings: ApiSettings, override val transport: Transport = Transport(apiSettings))
+    : AuthSession(apiSettings, transport) {
+    private var random = SecureRandom()
+    private var codeVerifier: String = ""
+    private val messageDigest = MessageDigest.getInstance("SHA-256") // "HmacSHA256")
+
+    init {
+        this.random = SecureRandom()
+    }
+
+    fun requestToken(body: Values): AuthToken {
+        val response = this.transport.request<AccessToken>(
+                HttpMethod.POST,
+                "/api/token",
+                mapOf(),
+                body)
+        val token = this.ok<AccessToken>(response)
+        this.authToken.setToken(token)
+        return this.authToken
+    }
+
+    override fun getToken(): AuthToken {
+        if (!this.isAuthenticated()) {
+            if (this.activeToken().refreshToken?.isNotEmpty()!!) {
+                val config = this.apiSettings.readConfig()
+                // fetch the token
+                this.requestToken(mapOf(
+                        "grant_type" to "refresh_token",
+                        "refresh_token" to this.activeToken().refreshToken,
+                        "client_id" to config["client_id"],
+                        "redirect_uri" to config["redirect_uri"]
+                ))
+            }
+        }
+        return this.activeToken()
+    }
+
+    /**
+     * Generate an OAuth2 authCode request URL
+     */
+    fun createAuthCodeRequestUrl(scope: String, state: String): String {
+        val bytes = this.secureRandom(32)
+        this.codeVerifier = this.sha256hash(bytes)
+        val config = this.apiSettings.readConfig()
+        val lookerUrl = config["looker_url"]
+        return addQueryParams("$lookerUrl/auth", mapOf(
+                "response_type" to "code",
+                "client_id" to config["client_id"],
+                "redirect_uri" to config["redirect_uri"],
+                "scope" to scope,
+                "state" to state,
+                "code_challenge_method" to "S256",
+                "code_challenge" to this.codeVerifier
+        ))
```

Comment:
This isn't quite right. The code verifier is the raw secure random number. The code challenge is that number hashed.

```suggestion
        this.codeVerifier = this.secureRandom(32)
        val codeChallenge = this.sha256hash(this.codeVerifier)
        val config = this.apiSettings.readConfig()
        val lookerUrl = config["looker_url"]
        return addQueryParams("$lookerUrl/auth", mapOf(
                "response_type" to "code",
                "client_id" to config["client_id"],
                "redirect_uri" to config["redirect_uri"],
                "scope" to scope,
                "state" to state,
                "code_challenge_method" to "S256",
                "code_challenge" to codeChallenge
        ))
```